source/Ship.cpp

```diff
@@ -2077,9 +2087,35 @@ bool Ship::CanLand() const
 
 
 
-bool Ship::CannotAct() const
+bool Ship::CannotAct(ActionType actionType) const
 {
-	return (zoom != 1.f || isDisabled || hyperspaceCount || pilotError || cloak);
+	bool canActCloaked = true;
+	if(cloak)
+		switch(actionType)
+		{
+			case ActionType::AFTERBURNER:
+				canActCloaked = attributes.Get("cloaked afterburner");
+				break;
+			case ActionType::BOARD:
+				canActCloaked = attributes.Get("cloaked boarding");
+				break;
+			case ActionType::COMMUNICATION:
+				canActCloaked = attributes.Get("cloaked communication");
+				break;
+			case ActionType::FIRE:
+				canActCloaked = attributes.Get("cloaked firing");
+				break;
+			case ActionType::PICKUP:
+				canActCloaked = attributes.Get("cloaked pickup");
+				break;
+			case ActionType::SCAN:
+				canActCloaked = attributes.Get("cloaked scanning");
+				break;
+		}
+	bool canSendHail = (actionType != ActionType::COMMUNICATION || crew);
+	return zoom != 1.f || isDisabled || hyperspaceCount || pilotError || !canSendHail
```

Comment:
I mean the first couple of conditions here have nothing to do with cloak so you could check them at the start of the function.

Reply:
right, now I get it